### ISSUE TITLE
Calling numpy.linalg.qr() on CuPy array fails

### ISSUE THREAD

**Issue by pentschev** [cupy/cupy#2011]
Code for reproduction and traceback:
```
In [1]: import cupy                                                                                                                                                                                         

In [2]: import numpy as np                                                                                                                                                                                  

In [3]: x = cupy.random.random((1000, 1000))                                                                                                                                                                

In [4]: np.linalg.qr(x)                                                                                                                                                                                     
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-4-4697d0fa0a15> in <module>
----> 1 np.linalg.qr(x)

~/.local/lib/python3.5/site-packages/numpy/linalg/linalg.py in qr(a, mode)
    901             raise ValueError("Unrecognized mode '%s'" % mode)
    902 
--> 903     a, wrap = _makearray(a)
    904     _assertRank2(a)
    905     m, n = a.shape

~/.local/lib/python3.5/site-packages/numpy/linalg/linalg.py in _makearray(a)
    115 
    116 def _makearray(a):
--> 117     new = asarray(a)
    118     wrap = getattr(a, "__array_prepare__", new.__array_wrap__)
    119     return new, wrap

~/.local/lib/python3.5/site-packages/numpy/core/numeric.py in asarray(a, dtype, order)
    536 
    537     """
--> 538     return array(a, dtype, copy=False, order=order)
    539 
    540 

ValueError: object __array__ method not producing an array
```

Running CuPy v6.0.0b2, NumPy master (commit https://github.com/numpy/numpy/commit/bf00ee5a431095f8877f23d1a6f46d158826eff1), CUDA 9.2, Ubuntu 16.04.5.

@mrocklin

**Comment by mrocklin** [comment]
Looking at the cupy implementation of `__array_function__` it looks like it may not work today for functions within modules

```python
    def __array_function__(self, func, types, args, kwargs):
        if not hasattr(cupy, func.__name__):
            return NotImplemented
        for t in types:
            if t not in _HANDLED_TYPES:
                return NotImplemented
        return getattr(cupy, func.__name__)(*args, **kwargs)
```

also cc @shoyer this may interest you

**Comment by shoyer** [comment]
You need to set the environment variable `NUMPY_EXPERIMENTAL_ARRAY_FUNCTION=1` before importing NumPy in order to enable `__array_function__` overloads at all.

**Comment by pentschev** [comment]
> You need to set the environment variable `NUMPY_EXPERIMENTAL_ARRAY_FUNCTION=1` before importing NumPy in order to enable `__array_function__` overloads at all.

This is only the case for NumPy v1.16, I'm using yesterday's master, which already includes your commit https://github.com/numpy/numpy/commit/96d179ce5bd0e3f75508f6afdfb2fb65170cc83c#diff-e5969453e399f2d32519d305b2582da9.

**Comment by pentschev** [comment]
> You need to set the environment variable `NUMPY_EXPERIMENTAL_ARRAY_FUNCTION=1` before importing NumPy in order to enable `__array_function__` overloads at all.

Sorry, you were right, I was indeed running v1.16, since pip installed it from its repository as a dependency to CuPy. Sorry about the confusion.

Nevertheless, it still fails, however, I'm now able to get the traceback when `__array_function__` is enabled, please see below:

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-7-4697d0fa0a15> in <module>
----> 1 np.linalg.qr(x)

~/.local/lib/python3.5/site-packages/numpy/core/overrides.py in public_api(*args, **kwargs)
    149             relevant_args = dispatcher(*args, **kwargs)
    150             return implement_array_function(
--> 151                 implementation, public_api, relevant_args, args, kwargs)
    152
    153         if module is not None:

TypeError: no implementation found for 'numpy.linalg.qr' on types that implement __array_function__: [<class 'cupy.core.core.ndarray'>]
```

**Comment by pentschev** [comment]
I'm suspecting the issue is related to how CuPy names its modules, it doesn't match the naming from NumPy:

```
In [1]: import cupy                                                                                                                                                                                                                                                           

In [2]: import numpy                                                                                                                                                                                                                                                          

In [3]: cupy.fft.fft.__module__                                                                                                                                                                                                                                               
Out[3]: 'cupy.fft.fft'

In [4]: numpy.fft.fft.__module__                                                                                                                                                                                                                                              
Out[4]: 'numpy.fft'

In [5]: cupy.linalg.qr.__module__                                                                                                                                                                                                                                             
Out[5]: 'cupy.linalg.decomposition'

In [6]: numpy.linalg.qr.__module__                                                                                                                                                                                                                                            
Out[6]: 'numpy.linalg'
```

**Comment by shoyer** [comment]
`getattr(cupy, func.__name__)` isn't going to work to find functions in submodules. You'll need to make use of `func.__module__`, too.

**Comment by mrocklin** [comment]
Right, this will probably require a slight modification to the current
implementation in cupy

**Comment by pentschev** [comment]
Alright, I finally understood how that was working. Thanks @mrocklin and @shoyer for pointing out where was the issue. A small change in implementation fixes the issue:

```python
    def __array_function__(self, func, types, args, kwargs):
        module_list = '.'.join(func.__module__.split('.')[1:])
        cupy_module = eval('cupy' if not module_list else 'cupy.' + module_list)
        if not hasattr(cupy_module, func.__name__):
            return NotImplemented
        for t in types:
            if t not in _HANDLED_TYPES:
                return NotImplemented
        return getattr(cupy_module, func.__name__)(*args, **kwargs)
```

I will soon open a PR for this.

**Comment by pentschev** [comment]
This issue was fixed in #2026.
